File: src/algorithms/bfs.js
// ─────────────────────────────────────────────────────────────────
// bfs.js — Breadth-First Search for minimum-stops route
// ─────────────────────────────────────────────────────────────────
// WHY BFS?
//   BFS explores stations level-by-level (1 stop away, then 2,
//   then 3...). So the FIRST time it reaches the destination,
//   that path is guaranteed to have the fewest stops.
//   Dijkstra would be needed for shortest TIME or lowest FARE,
//   but for fewest stops, BFS is perfect.
//
// HOW IT WORKS:
//   1. Start with a queue containing just [source].
//   2. Take the first path from the queue.
//   3. If its last station is the destination → done!
//   4. Otherwise, add each unvisited neighbour as a new path.
//   5. Repeat until found or queue is empty (no route).
//
// RETURNS:
//   - Array of station names (the path), e.g. ["A", "B", "C"]
//   - null if no route exists
// ─────────────────────────────────────────────────────────────────

function bfs(graph, source, destination) {

    const queue = [[source]];
    const visited = new Set();

    visited.add(source);

    while (queue.length > 0) {

        const path = queue.shift();
        const current = path[path.length - 1];

        // Found the destination — return the path
        if (current === destination) {
            return path;
        }

        // Explore all neighbours of the current station
        const neighbours = graph.getNeighbours(current);

        for (const edge of neighbours) {

            const next = edge.station;

            if (!visited.has(next)) {
                visited.add(next);
                queue.push([...path, next]);
            }
        }
    }

    // Queue is empty and destination was never reached
    return null;
}

module.exports = bfs;

File: src/algorithms/bfs.cpp
// ─────────────────────────────────────────────────────────────────
// bfs.cpp — Breadth-First Search for minimum-stops route
// ─────────────────────────────────────────────────────────────────
// WHY BFS?
//   BFS explores stations level-by-level (1 stop away, then 2,
//   then 3...). The FIRST time it reaches the destination, that
//   path is guaranteed to have the fewest stops.
//
// HOW IT WORKS:
//   1. Start with a queue containing just [source].
//   2. Take the first path from the queue.
//   3. If its last station is the destination → done!
//   4. Otherwise, push each unvisited neighbour as a new path.
//   5. Repeat until found or queue is empty (no route).
//
// RETURNS:
//   - Vector of station names (the path), e.g. {"A","B","C"}
//   - Empty vector if no route exists
// ─────────────────────────────────────────────────────────────────
#include "algorithms/bfs.h"
#include <queue>
#include <unordered_set>

using namespace std;

vector<string> bfs(const Graph& graph,
                              const string& source,
                              const string& destination) {

    // Queue of paths (each path is a vector of station names)
    queue<vector<string>> q;
    unordered_set<string> visited;

    q.push({source});
    visited.insert(source);

    while (!q.empty()) {
        vector<string> path = q.front();
        q.pop();

        const string& current = path.back();

        // Found the destination — return this path
        if (current == destination) {
            return path;
        }

        // Explore all neighbours of the current station
        for (const Edge& edge : graph.getNeighbours(current)) {
            const string& next = edge.station;

            if (visited.find(next) == visited.end()) {
                visited.insert(next);
                vector<string> newPath = path;
                newPath.push_back(next);
                q.push(std::move(newPath));
            }
        }
    }

    // Queue empty — no route found
    return {};
}
